src/envoy/http/mixer/filter.cc

```diff
@@ -164,26 +165,74 @@ FilterTrailersStatus Filter::decodeTrailers(HeaderMap& trailers) {
   return FilterTrailersStatus::Continue;
 }
 
+void Filter::UpdateHeaders(
+    HeaderMap* headers, const ::google::protobuf::RepeatedPtrField<
+                            ::istio::mixer::v1::HeaderOperation>& operations) {
+  for (auto const iter : operations) {
+    switch (iter.operation()) {
+      case ::istio::mixer::v1::HeaderOperation_Operation_REPLACE:
+        headers->remove(LowerCaseString(iter.name()));
+        headers->addCopy(LowerCaseString(iter.name()), iter.value());
+        break;
+      case ::istio::mixer::v1::HeaderOperation_Operation_REMOVE:
+        headers->remove(LowerCaseString(iter.name()));
+        break;
+      case ::istio::mixer::v1::HeaderOperation_Operation_APPEND:
+        headers->addCopy(LowerCaseString(iter.name()), iter.value());
+        break;
+      default:
+        PANIC("unreachable header operation");
+    }
+  }
+}
+
+FilterHeadersStatus Filter::encodeHeaders(HeaderMap& headers, bool) {
+  ENVOY_LOG(debug, "Called Mixer::Filter : {} {}", __func__, state_);
+  ASSERT(state_ == Complete || state_ == Responded);
+  if (state_ == Complete) {
+    // handle response header operations
+    UpdateHeaders(&headers, route_directive_.update_response_headers());
+  }
+  return FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus Filter::encodeData(Buffer::Instance&, bool) {
+  ENVOY_LOG(debug, "Called Mixer::Filter : {}", __func__);
+  if (state_ == Calling) {
```

Comment:
Same as encodeHeaders,  we can assert the state_ as Complete or Respond.

Reply:
done